internal/cluster/cluster_operations.go

```diff
@@ -359,70 +357,86 @@ func tryConvertToClusterDatastoreSpec(object map[string]interface{}, clusterName
 	return result, nil
 }
 
-func FlattenCluster(ctx context.Context, clusterObj *models.Cluster, apiClient *client.VcfClient) (*map[string]interface{}, error) {
+func FlattenCluster(ctx context.Context, clusterObj *vcf.Cluster, apiClient *vcf.ClientWithResponses) (*map[string]interface{}, error) {
 	result := make(map[string]interface{})
 	if clusterObj == nil {
 		return &result, nil
 	}
 
-	result["id"] = clusterObj.ID
+	result["id"] = clusterObj.Id
 	result["name"] = clusterObj.Name
 	result["primary_datastore_name"] = clusterObj.PrimaryDatastoreName
 	result["primary_datastore_type"] = clusterObj.PrimaryDatastoreType
 	result["is_default"] = clusterObj.IsDefault
 	result["is_stretched"] = clusterObj.IsStretched
 
-	flattenedVdsSpecs := getFlattenedVdsSpecsForRefs(clusterObj.VdsSpecs)
-	result["vds"] = flattenedVdsSpecs
+	if clusterObj.VdsSpecs != nil {
+		flattenedVdsSpecs := getFlattenedVdsSpecsForRefs(*clusterObj.VdsSpecs)
+		result["vds"] = flattenedVdsSpecs
+	}
 
-	flattenedHostSpecs, err := getFlattenedHostSpecsForRefs(ctx, clusterObj.Hosts, apiClient)
-	if err != nil {
-		return nil, err
+	if clusterObj.Hosts != nil {
+		flattenedHostSpecs, err := getFlattenedHostSpecsForRefs(ctx, *clusterObj.Hosts, apiClient)
+		if err != nil {
+			return nil, err
+		}
+		result["host"] = flattenedHostSpecs
 	}
-	result["host"] = flattenedHostSpecs
 
 	return &result, nil
 }
 
-func ImportCluster(ctx context.Context, data *schema.ResourceData, apiClient *client.VcfClient,
+func ImportCluster(ctx context.Context, data *schema.ResourceData, apiClient *vcf.ClientWithResponses,
 	clusterId string) ([]*schema.ResourceData, error) {
-	getClusterParams := clusters.NewGetClusterParamsWithContext(ctx).
-		WithTimeout(constants.DefaultVcfApiCallTimeout)
-	getClusterParams.ID = clusterId
-	clusterResult, err := apiClient.Clusters.GetCluster(getClusterParams)
+	clusterRes, err := apiClient.GetClusterWithResponse(ctx, clusterId)
 	if err != nil {
 		return nil, err
 	}
-	clusterObj := clusterResult.Payload
+	if clusterRes.StatusCode() != 200 {
```

Comment:
Again, I'd like to see all these checks being moved to a utility

Reply:
I did a utility for the error cases but not for the success codes, because some APIs respond with 200, some with 201, 202 etc.

I could replace this with something like `someutility.CheckStatus(resp, 200) but it would still be necessary to specify the correct code for the particular operation.

Reply:
for sucess then you can check that the code is above 199 and below 300

Reply:
someutility.CheckStatus(resp, 200) still leaks http logic

Reply:
More like a `utils.IsOK(resp)`?

```go
package utils

import "net/http"

func IsOK(resp *http.Response) bool {
    return resp.StatusCode == http.StatusOK
}

func IsNotFound(resp *http.Response) bool {
    return resp.StatusCode == http.StatusNotFound
}
```

Reply:
Take a look `3cb3ae4b31d21b5814cca92810ea660fc715b783`

I added a new utility method that solves the problem without looking at the actual status codes.
It leverages the fact that if the operation is successful its body can be unmarshalled into the expected return type, and into vcf.Error otherwise.

```
// GetResponseAs attempts to parse the response body into the provided type
// If it fails it attempts to parse it as a vcf.Error
func GetResponseAs[T interface{}](body []byte) (*T, *vcf.Error) {
	var resp T
	if json.Unmarshal(body, &resp) != nil {
		return nil, getError(body)
	}

	return &resp, nil
}

// getError when the API responds with an error code the response is unmarshalled into the appropriate field for that code
// all error code fields are of type *vcf.Error and only one can be != nil at any time
// if the status code is an error code the body is always *vcf.Error
//
// use this method if you are not interested in the error code but only in the error itself.
func getError(body []byte) *vcf.Error {
	var dest vcf.Error
	if json.Unmarshal(body, &dest) != nil {
		return nil
	}

	return &dest
}
```

And here's an example of how its used

```
task, vcfErr := api_client.GetResponseAs[vcf.Task](res.Body)
if vcfErr != nil {
	api_client.LogError(vcfErr)
	return errors.New(*vcfErr.Message)
}
if err := sddcClient.WaitForTask(ctx, *task.Id); err != nil {
	return err
}
```

Reply:
Love it!